python_modules/dagster/dagster/core/asset_defs/asset_group.py

```diff
@@ -364,6 +359,51 @@ def from_package_name(
             package_module, resource_defs=resource_defs, executor_def=executor_def
         )
 
+    @staticmethod
+    def from_modules(
+        modules: Sequence[ModuleType],
+        resource_defs: Optional[Mapping[str, ResourceDefinition]] = None,
+        executor_def: Optional[ExecutorDefinition] = None,
+    ) -> "AssetGroup":
+        """
+        Constructs an AssetGroup that includes all asset definitions in the given module.
+
+        Args:
+            modules (Sequence[ModuleType]): The Python modules to look for assets inside.
+            resource_defs (Optional[Mapping[str, ResourceDefinition]]): A dictionary of resource
+                definitions to include on the returned asset group.
+            executor_def (Optional[ExecutorDefinition]): An executor to include on the returned
+                asset group.
+
+        Returns:
+            AssetGroup: An asset group with all the assets defined in the given modules.
+        """
+        assets: Set[AssetsDefinition] = set()
+        source_assets: Set[SourceAsset] = set()
+        for module in modules:
+            _add_assets_in_module(module, assets, source_assets)
+
+        return AssetGroup(
+            assets=list(assets),
+            source_assets=list(source_assets),
+            resource_defs=resource_defs,
+            executor_def=executor_def,
+        )
+
+
+def _add_assets_in_module(
+    module: ModuleType, assets: Set[AssetsDefinition], source_assets: Set[SourceAsset]
+) -> None:
+    """
+    Finds assets in the given module and adds them to the given sets of assets and source assets.
+    """
+    for attr in dir(module):
+        value = getattr(module, attr)
+        if isinstance(value, AssetsDefinition):
+            assets.add(value)
+        if isinstance(value, SourceAsset):
+            source_assets.add(value)
```

Comment:
nit, but it's kinda weird to me that we would mutate state in this helper function instead of just returning the respective lists of assets and source assets, and adding them in the method.

Reply:
oy yeah - I was trying to avoid creating unnecessary intermediate lists. I wish a function could be a double generator. maybe that's premature optimization though - changed it to return a tuple of lists.